File: test/language/pubsub_types/cpp/TestPubsub.cpp
#include "TestPubsub.h"
#include "zserio/PubsubException.h"

namespace pubsub_types
{

void TestPubsub::publish(const std::string& topic, const std::vector<uint8_t>& data, void* context)
{
    if (context != nullptr)
        static_cast<Context*>(context)->seenByPubsub = true;

    for (const auto& subscription : m_subscriptions)
    {
        if (subscription.second.topic == topic)
            subscription.second.callback(topic, data);
    }
}

zserio::IPubsub::SubscriptionId TestPubsub::subscribe(const std::string& topic, const OnTopic& callback, void *context)
{
    if (context != nullptr)
        static_cast<Context*>(context)->seenByPubsub = true;

    m_subscriptions.emplace(m_numIds, Subscription{topic, callback});
    return m_numIds++;
}

void TestPubsub::unsubscribe(SubscriptionId id)
{
    const auto found = m_subscriptions.find(id);
    if (found == m_subscriptions.end())
        throw zserio::PubsubException("TestPubsub: Invalid subscription ID '" + std::to_string(id) + "'!");
    m_subscriptions.erase(found);
}

} // namespace pubsub_types

File: test/language/pubsub_types/python/TestPubsub.py
import zserio


class TestPubsub(zserio.PubsubInterface):
    def __init__(self):
        self._subscriptions = {}
        self._numIds = 0

    def publish(self, topic, data, context=None):
        if context:
            context.seenByPubsub = True

        for subscription in self._subscriptions.values():
            if subscription["topic"] == topic:
                subscription["callback"](topic, data)

    def subscribe(self, topic, callback, context=None):
        if context:
            context.seenByPubsub = True

        subscriptionId = self._numIds
        self._numIds += 1
        self._subscriptions[subscriptionId] = {"topic": topic, "callback": callback}
        return subscriptionId

    def unsubscribe(self, subscription_id):
        if not subscription_id in self._subscriptions:
            raise zserio.PubsubException(f"TestPubsub: Invalid subscription ID '{subscription_id}'!")

        self._subscriptions.pop(subscription_id)


class TestPubsubContext:
    def __init__(self):
        self.seenByPubsub = False
